lib/loadEnv.test.js

```diff
@@ -0,0 +1,59 @@
+'use strict';
+
+const chai = require('chai');
+const path = require('path');
+const sinon = require('sinon');
+const overrideEnv = require('process-utils/override-env');
+const overrideCwd = require('process-utils/override-cwd');
+const loadEnv = require('./loadEnv');
+const writeFileSync = require('./utils/fs/writeFileSync');
+const { createTmpDir } = require('../test/utils/fs');
+const dotenv = require('dotenv');
+const ServerlessError = require('./classes/Error').ServerlessError;
+
+chai.use(require('chai-as-promised'));
+const expect = require('chai').expect;
+
+describe('loadEnv', () => {
+  let restoreCwd;
+  let restoreEnv;
+
+  beforeEach(() => {
+    const tmpDir = createTmpDir();
```

Comment:
This is a place that I would love to get your feedback @medikoo - I was trying to just setup `.env` files once in `before`, however, after first test, the files were no longer accessible, so I switched to `beforeEach`/`afterEach` combo that I'm not super happy about due to unnecessary creation of `.env` files before each test. Do you have any suggestions on how to properly set it up so these files setup in `before` will be visible to all tests?

Reply:
@pgrzesik As we're preparing it as standalone module (that takes `stage`), the unit test may simply put needed `.env` files to `process.cwd()`, and confirm that after running util expected env vars are seen on `process.env`

Important notes:
- `process.env` is reset of each test file, therefore we don't have to worry about side effects (it's ensured by: [`restore-env`](https://github.com/serverless/test/blob/master/docs/setup/restore-env.md) Mocha extension we have loaded
- `process.cwd()` is set to home dir, which is set to temporary dir (again ensured by Mocha extensions: [`mock-cwd`](https://github.com/serverless/test/blob/master/docs/setup/mock-cwd.md) and [`mock-homedir`](https://github.com/serverless/test/blob/master/docs/setup/mock-homedir.md) - note that homedir is cleanup up for other test files, so we also do not need to take care of cleaning it.

Additionally if you'd like to have `process.env` or `process.cwd` isolated for specific test runs (so there are no side effects for following tests in same test file). check: [`process-utils/override-env`](https://github.com/medikoo/process-utils/blob/master/docs/override-env.md) an [`process-utils/override-cwd`](https://github.com/medikoo/process-utils/blob/master/docs/override-cwd.md)

When having it isolated, and tested as above with unit test, not sure if it's worth testing that resolution works in context of Serverless instance. We might have some one sanity test, and such can be constructed as:

1. Setup a chosen fixture (like here: https://github.com/serverless/serverless/blob/c4c57b08a08c14b2535d6158100636e60c6ee01d/lib/plugins/aws/package/compile/functions/index.test.js#L2931)
2. Then in returned `servicePath` setup needed `.env` files
3. Invoke `runServerless` run on fixture (as here: https://github.com/serverless/serverless/blob/c4c57b08a08c14b2535d6158100636e60c6ee01d/lib/plugins/aws/package/compile/functions/index.test.js#L2933-L2936)
4. Confirm on the outcome

--- 

If that doesn't answer your questions, let me know. I know I've provided an indirect answer, but I hope that above information clarifies how to setup test efficiently

Reply:
I've posted above comment, without looking into test file, which looks very good imo.

I believe you don't need to create tmp dir, you may just put files in `process.cwd()` (once) as is, and all that you need to reset is `process.env` (which you already do with `process-utils/override-env`).

 `.env` files should definitely not be removed between single tests, so you should be able to setup them once in `before` (they will however be cleaned after whole test file runs)